### PR TITLE
Add any/overlap array lookups to Elasticsearch and memory adapters

### LINKED ISSUE #951: 3.6.2: Adapter parity for query lookups
## Summary

Bring the memory adapter to lookup parity with the SQLAlchemy adapter. Where the gap is genuine — a lookup that the memory adapter cannot semantically reproduce — raise `NotImplementedError` loudly with a clear message rather than silently diverging. Audit the Elasticsearch adapter against the same matrix.

## Problem

Memory and SQLAlchemy adapters have different lookup support today. The outbox audit surfaced specific gaps:

- `__lt`, `__lte`, `__gt`, `__gte`: SQLAlchemy supports them; memory adapter handles only equality and `__in` ([memory.py L357-369](../blob/main/src/protean/adapters/repository/memory.py#L357-L369)).
- `__isnull`: nowhere supported until #942 adds it. After #942, SQLAlchemy and memory both have it.
- `F()` column references: SQLAlchemy native; memory implements via dict-attribute lookup; ES via Painless script.

This causes three failure modes:

1. **Silent divergence.** Code paths that filter using `__lt` work on SQLAlchemy but raise `KeyError` or return wrong results on memory. Tests that use the autouse `test_domain` fixture (which uses memory) pass while production fails.
2. **Test untrustworthiness.** Tests written against memory don't verify the SQL semantics that production runs.
3. **Adapter authoring confusion.** No clear list of which lookups are "must support" vs "can opt out."

### Lookup taxonomy

Every Protean lookup falls into one of three categories.

**Category 1: Trivially universal.** The lookup is a basic comparison and every store can express it natively.

- `exact`, `iexact`, `contains`, `gt`, `gte`, `lt`, `lte`, `in`, `isnull`

These should work on every adapter.

**Category 2: Adapter-specific implementation, same semantics.** The lookup has a clear meaning but adapters implement it differently.

- `F()` column reference: SQLAlchemy native; memory dict lookup; ES Painless script.
- Regex match: SQLAlchemy/Postgres uses POSIX regex; ES uses regexp query; memory uses `re` module.

These should work on every adapter, with adapter-specific implementation cost noted in docs.

**Category 3: Genuinely backend-specific.** No clear cross-adapter semantics.

- Full-text search (`__search`): meaningful in ES, not really in memory or SQLAlchemy without extensions.
- Geospatial lookups.
- Vector / embedding similarity.

These should be explicit opt-ins, not present on every adapter.

### Memory adapter today

The dict-based filter dispatcher recognizes equality and `__in` but not lookup suffixes generally. To extend it, the dispatcher needs to recognize lookup suffixes (`__lt`, `__gte`, etc.) and apply the right Python operation.

### Loud failures

For lookups that genuinely don't apply to a backend (memory adapter receiving a `__search` query, e.g.), the adapter should raise:

```python
raise NotImplementedError(
    f"Lookup '{lookup}' is not supported on the memory adapter. "
    f"This lookup requires a backend with full-text search capability "
    f"(e.g., Elasticsearch). Either change the test to use a supported "
    f"backend or replace the lookup with an equivalent expression."
)
```

The error message names the missing capability, points to a backend that has it, and offers an actionable next step.

### A. Memory adapter lookup parity

Bring the memory adapter to category-1 + category-2 parity:

```python
# src/protean/adapters/repository/memory.py

LOOKUP_HANDLERS = {
    "exact":   lambda v, target: v == target,
    "iexact":  lambda v, target: str(v).lower() == str(target).lower(),
    "contains": lambda v, target: target in v if v is not None else False,
    "gt":      lambda v, target: v is not None and v > target,
    "gte":     lambda v, target: v is not None and v >= target,
    "lt":      lambda v, target: v is not None and v < target,
    "lte":     lambda v, target: v is not None and v <= target,
    "in":      lambda v, target: v in target,
    "isnull":  lambda v, target: (v is None) == bool(target),
}

def _apply_lookup(self, item, field_name, lookup, target):
    handler = LOOKUP_HANDLERS.get(lookup)
    if handler is None:
        raise NotImplementedError(
            f"Lookup '{lookup}' is not supported on the memory adapter. "
            f"See docs/concepts/query-api.md#lookup-matrix for the supported set."
        )
    value = self._resolve_value(item, field_name)
    return handler(value, target)
```

NULL handling: `gt`/`gte`/`lt`/`lte` return False when the value is None. This matches SQL semantics where comparisons against NULL produce NULL/UNKNOWN, treated as not-true in WHERE clauses.

`isnull` is the explicit way to test for NULL; the comparison lookups never return True when one side is None.

### B. F() resolution in memory adapter

When the right-hand side of a lookup is an `F()`, resolve it against the candidate item:

```python
def _resolve_value(self, item, value):
    if isinstance(value, F):
        return item.get(value.name)
    return value
```

Already covered partly by epic #941 sub-issue #945 for the F() introduction; this sub-issue ensures the memory adapter implements the resolver consistently with the contract.

### C. Conformance test suite

A new test module `tests/port/test_lookup_parity.py` runs the same lookup tests against every adapter:

```python
@pytest.mark.dao_contract
@pytest.mark.parametrize("adapter", ALL_ADAPTERS_AS_FIXTURES)
class TestLookupParity:
    def test_lt_excludes_equal(self, adapter):
        adapter.populate([1, 2, 3])
        result = adapter.query.filter(value__lt=2).all()
        assert result.total == 1

    def test_isnull_true_finds_nones(self, adapter):
        adapter.populate([1, None, 3])
        result = adapter.query.filter(value__isnull=True).all()
        assert result.total == 1

    # ... one test per category-1 and category-2 lookup
```

Every category-1 and category-2 lookup gets a parametrized test across all adapters. Test failures pin the regression to a specific adapter / lookup combination.

### D. Loud-failure documentation

For category-3 lookups, a matrix in `docs/concepts/query-api.md`:

| Lookup         | Memory | SQLAlchemy | Elasticsearch |
|----------------|--------|------------|---------------|
| `exact`        | ✅     | ✅         | ✅            |
| `iexact`       | ✅     | ✅         | ✅            |
| `contains`     | ✅     | ✅         | ✅            |
| `gt` / `gte`   | ✅     | ✅         | ✅            |
| `lt` / `lte`   | ✅     | ✅         | ✅            |
| `in`           | ✅     | ✅         | ✅            |
| `isnull`       | ✅     | ✅         | ✅            |
| `F()` ref      | ✅     | ✅         | ✅ (script)   |
| `search`       | ❌     | ❌         | ✅            |
| `match_phrase` | ❌     | ❌         | ✅            |

When a test uses `__search` and runs against the memory adapter, the `NotImplementedError` carries the matrix-row guidance.

### E. ES adapter audit

Audit the Elasticsearch adapter against the same matrix. Document any category-1 lookups it doesn't support and bring it to parity for the universal subset. Category-2 lookups get adapter-specific rendering documented in the adapter's section of the query-api page.

## Alternatives considered

- **Convert silent divergences to "best effort" emulation.** Rejected — silent best-effort is worse than loud failure. If `__lt` on memory falls back to "ignore the predicate" the test passes when it shouldn't.
- **Drop the memory adapter entirely; require SQLAlchemy in tests.** Rejected — the memory adapter is fast enough for unit tests that the migration cost would be huge. Bringing it to parity is the smaller change.
- **Adapter capability reporting** (each adapter exposes `supported_lookups`) and let consumers degrade gracefully. Rejected — pushes the parity question onto consumers; complicates every callsite. Loud failure at adapter level is cleaner.

## Backward compatibility

Pure addition of lookup support; no removed behavior. Tests that previously passed because memory silently ignored a `__lt` predicate may now fail — that is a correctness improvement, but it is a behavioral change for those tests. Migration: fix the tests to assert what they actually mean, or use the SQLAlchemy adapter for tests that need it.

## Test plan

- `tests/port/test_lookup_parity.py` — parametrized lookup tests across all adapters (the primary deliverable).
- `tests/adapters/repository/memory/test_lookups.py` — memory-adapter-specific edge cases (NULL handling, type coercion, mixed-type comparisons).
- `tests/adapters/repository/elasticsearch/test_lookups.py` — ES-adapter-specific edge cases.
- Cross-adapter outbox tests — outbox repository tests run against memory and SQLAlchemy, asserting identical behavior.

## Acceptance criteria

- [ ] Memory adapter supports all category-1 lookups (`exact`, `iexact`, `contains`, `gt`, `gte`, `lt`, `lte`, `in`, `isnull`).
- [ ] Memory adapter supports category-2 lookups including `F()` resolution.
- [ ] Memory adapter raises `NotImplementedError` with actionable message for category-3 lookups.
- [ ] Lookup parity matrix documented in `docs/concepts/query-api.md`.
- [ ] Conformance suite (`tests/port/test_lookup_parity.py`) runs the same lookup tests against memory, SQLAlchemy, ES and passes.
- [ ] ES adapter audited; gaps either fixed or documented as opt-out with the same loud-failure pattern.

### PR DESCRIPTION
**Pull request by subhashb**
## Summary

Brings the `any` and `overlap` array-membership lookups to parity across adapters (issue #951).

- **Elasticsearch**: adds `any` and `overlap`. ES fields are natively multivalued, so both render as a `terms` query. `Any` subclasses `In` (inheriting the `.keyword` handling); `Overlap` subclasses `Any`. A scalar `any` target is normalised to a single-element list.
- **In-memory**: adds `overlap` (subclasses `Any`). Also hoists the coerced target out of the per-record generator. Membership stays list-based (not a set) so unhashable elements (e.g. `List(content_type=dict)`, serialized value objects) compare by equality rather than raising.
- **SQLAlchemy** already supported both via native array operators (`= ANY(...)` / `&&`); verified on PostgreSQL.

Adds a per-adapter **lookup support matrix** to `docs/concepts/internals/query-system.md`, linked from the retrieve-aggregates guide. An unsupported lookup raises `NotImplementedError` rather than returning wrong results.

## Lookup matrix

| Lookup | Memory | SQLAlchemy | Elasticsearch |
|---|:--:|:--:|:--:|
| 12 core lookups (`exact`…`isnull`) | ✅ | ✅ | ✅ |
| `any` / `overlap` (array) | ✅ | ✅ (native array) | ✅ |
| `F()` column ref | ✅ | ✅ | ❌ |

## Test plan

- [x] Core tests pass (9445)
- [x] Array lookups: ES live end-to-end + in-memory (incl. unhashable-element guard)
- [x] Generic filtering conformance green on memory / sqlite / postgresql / elasticsearch
- [x] 100% patch coverage

## Notes

- Cross-adapter `isnull` conformance already exists in `tests/repository/test_count_and_isnull.py`, so no duplicate was added here.

Closes #951

### CODE REVIEW

File: src/protean/adapters/repository/elasticsearch.py [hunk 1]
```diff
@@ -1148,6 +1148,36 @@ def as_expression(self):
...
+    ``terms`` query as :class:`In`; the only difference is that ``any`` also
+    accepts a scalar target, normalised here to a single-element list.
+    """
+
+    lookup_name = "any"
+
+    def process_target(self):
+        if not isinstance(self.target, (list, tuple)):
+            self.target = [self.target]
+        return super().process_target()
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/adapters/repository/elasticsearch.py:1169]: Fixed — a scalar target now goes through `DefaultLookup.process_target` first (rejecting `F()` with `NotImplementedError`, stringifying `UUID`) before being wrapped for the `terms` query. Added tests for both the scalar-`F` raise and scalar-`UUID` stringification.

File: docs/concepts/internals/query-system.md [hunk 2]
```diff
@@ -282,6 +282,37 @@ Because `F` resolution lives in `process_target`, it composes with every
...
+| `overlap` (array) | ✅     | ✅ (native array) | ✅            |
+| `F()` column ref  | ✅     | ✅                | ❌            |
+
+The twelve lookups above `any` are universal and exercised by the cross-adapter
+conformance suites under `tests/adapters/repository/generic/` and
+`tests/repository/`. The array lookups `any` and `overlap` match documents whose
+array field shares an element with the given values: SQLAlchemy maps them to native array operators
+(`= ANY(...)` and `&&`, so they need a backend with array columns such as
+PostgreSQL); Elasticsearch renders both as a `terms` query (its fields are
+natively multivalued); the in-memory store evaluates a set intersection.
```
[the remark this answers is not in the document]
Comment (subhashb) [reply docs/concepts/internals/query-system.md]: Reworded — the matrix now says the in-memory store "compares elements by equality (not hashing, so unhashable items such as dicts are supported)" instead of "set intersection".

File: src/protean/adapters/repository/memory.py [hunk 3]
```diff
@@ -807,7 +807,24 @@ def evaluate(self) -> bool:
...
+
+@MemoryProvider.register_lookup
+class Overlap(Any):
+    """Array overlap query.
+
+    Matches when the source list shares at least one element with the target
+    list. For the in-memory store this is the same set-intersection test as
+    :class:`Any`; the distinct name keeps parity with the SQLAlchemy adapter,
+    where ``overlap`` maps to the native array ``&&`` operator.
+    """
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/adapters/repository/memory.py:827]: Reworded the docstring to describe equality-based membership and note that elements are compared by equality (not hashing) so unhashable items like dicts are supported.
